subsys/dfu/img_util/flash_block.c

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2017 Nordic Semiconductor ASA
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define SYS_LOG_DOMAIN "fota/flash_block"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_FOTA_LEVEL
+#include <logging/sys_log.h>
+
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <flash.h>
+#include "boot/mcuboot_constraints.h"
+#include "img_util/flash_block.h"
+
+#if (CONFIG_IMG_BLOCK_BUF_SIZE % FLASH_ALIGN != 0)
+#error CONFIG_IMG_BLOCK_BUF_SIZE is not multiple of FLASH_ALIGN
+#endif
+
+static bool flash_verify(struct device *dev, off_t offset,
+			 u8_t *data, int len)
+{
+	u32_t temp;
+	int size;
+	int rc;
+
+	while (len) {
+		size = (len >= sizeof(temp)) ? sizeof(temp) : len;
+		rc = flash_read(dev, offset, &temp, size);
+		if (rc) {
+			SYS_LOG_ERR("flash_read error %d addr=0x%08x",
+				    rc, offset);
+			break;
+		}
+
+		if (memcmp(data, &temp, size)) {
+			SYS_LOG_ERR("addr=0x%08x VERIFY FAIL", offset);
+			break;
+		}
+		len -= size;
+		offset += size;
+		data += size;
+	}
+
+	return (len == 0) ? true : false;
+}
+
+/* buffer data into block writes */
+static int flash_block_write(struct flash_img_contex *cb, off_t offset,
+			     u8_t *data, int len, bool finished)
+{
+	int processed = 0;
+	int rc = 0;
+
+	while ((len - processed) >
+	       (CONFIG_IMG_BLOCK_BUF_SIZE - cb->buf_bytes)) {
+		memcpy(cb->buf + cb->buf_bytes, data + processed,
+		       (CONFIG_IMG_BLOCK_BUF_SIZE - cb->buf_bytes));
```

Comment:
Why can't we just write directly from 'data'? Or is this buffer somehow coming in misaligned (some drivers won't handle that).